From e9d65a3d8378b6d3fb00ac43ad89e6c4c7b298ac Mon Sep 17 00:00:00 2001
From: Brett Mastbergen <bmastbergen@ciq.com>
Date: Wed, 14 Jan 2026 10:18:34 -0500
Subject: [PATCH] tls: wait for pending async decryptions if tls_strp_msg_hold
 fails

jira VULN-162729
cve CVE-2025-40176
commit-author Sabrina Dubroca <sd@queasysnail.net>
commit b8a6ff84abbcbbc445463de58704686011edc8e1

Async decryption calls tls_strp_msg_hold to create a clone of the
input skb to hold references to the memory it uses. If we fail to
allocate that clone, proceeding with async decryption can lead to
various issues (UAF on the skb, writing into userspace memory after
the recv() call has returned).

In this case, wait for all pending decryption requests.

Fixes: 84c61fe1a75b ("tls: rx: do not use the standard strparser")
	Reported-by: Jann Horn <jannh@google.com>
	Signed-off-by: Sabrina Dubroca <sd@queasysnail.net>
Link: https://patch.msgid.link/b9fe61dcc07dab15da9b35cf4c7d86382a98caf2.1760432043.git.sd@queasysnail.net
	Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit b8a6ff84abbcbbc445463de58704686011edc8e1)
	Signed-off-by: Brett Mastbergen <bmastbergen@ciq.com>
---
 net/tls/tls_sw.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/net/tls/tls_sw.c b/net/tls/tls_sw.c
index cd22968bf831..a58b81d60010 100644
--- a/net/tls/tls_sw.c
+++ b/net/tls/tls_sw.c
@@ -1632,8 +1632,10 @@ static int tls_decrypt_sg(struct sock *sk, struct iov_iter *out_iov,
 
 	if (unlikely(darg->async)) {
 		err = tls_strp_msg_hold(&ctx->strp, &ctx->async_hold);
-		if (err)
-			__skb_queue_tail(&ctx->async_hold, darg->skb);
+		if (err) {
+			err = tls_decrypt_async_wait(ctx);
+			darg->async = false;
+		}
 		return err;
 	}
 
-- 
2.47.3

